src/main.zig

```diff
@@ -136,14 +144,49 @@ pub fn serialize(comptime T: type, data: T, l: *ArrayList(u8)) !void {
                 else => return error.UnSupportedPointerType,
             }
         },
-        .Struct => {
+        .Struct => |struc| {
             // First pass, accumulate the fixed sizes
             comptime var var_start = 0;
-            inline for (info.Struct.fields) |field| {
-                if (@typeInfo(field.type) == .Int or @typeInfo(field.type) == .Bool) {
-                    var_start += @sizeOf(field.type);
-                } else {
-                    var_start += 4;
+            comptime var is_stable_container: bool = struc.fields.len > 0;
+            comptime var last_field_index: usize = 0;
+            inline for (struc.fields, 0..) |field, i| {
+                last_field_index = i;
+                switch (@typeInfo(field.type)) {
+                    .Int, .Bool => {
+                        var_start += @sizeOf(field.type);
+                        is_stable_container = false;
+                    },
+                    else => {
+                        var_start += 4;
+                        is_stable_container = is_stable_container and (@typeInfo(field.type) == .Optional or i + 1 == info.Struct.fields.len);
+                    },
+                }
+            }
+            if (is_stable_container) {
+                // check that the type of the last field is
+                // the one we are looking for.
+                const last_field = struc.fields[last_field_index];
+                if (std.mem.indexOf(u8, @typeName(last_field.type), "StableContainerFiller") != null) {
```

Comment:
the check is really loose here, which is a huge drawback with this method. Nonetheless, compilation would fail if that type didn't have a `max_size` constant, so if someone wants to create their own type, good for them.

Come to think of it, relaxing the check on the name might even be simpler: anyone is free to declare their own type, and `StabeContainerFiller` is only provided for convenience.